python/tank/descriptor/io_descriptor/git.py

```diff
@@ -43,17 +56,109 @@ def __init__(self, descriptor_dict):
+        :raises: TankGitError on git failure
         """
+        # ensure *parent* folder exists
+        parent_folder = os.path.dirname(target)
+        filesystem.ensure_folder_exists(parent_folder)
+
+        # first probe to check that git exists in our PATH
+        log.debug("Checking that git exists and can be executed...")
+        try:
+            output = subprocess_check_output(["git", "--version"])
+        except:
+            raise TankGitError(
+                "Cannot execute the 'git' command. Please make sure that git is "
+                "installed on your system and that the git executable has been added to the PATH."
+            )
+        log.debug("Git installed: %s" % output)
+
         # Note: git doesn't like paths in single quotes when running on
-        # windows - it also prefers to use forward slashes!
-        log.debug("Git Cloning %r into %s" % (self, target_path))
-        execute_git_command("clone -q \"%s\" \"%s\"" % (self._sanitized_repo_path, target_path))
+        # windows - it also prefers to use forward slashes
+        #
+        # note that we use os.system here to allow for git to pop up (in a terminal
+        # if necessary) authentication prompting. This DOES NOT seem to be possible
+        # with subprocess.
+        #
+        log.debug("Git Cloning %r into %s" % (self, target))
+        cmd = "git clone -q \"%s\" \"%s\"" % (self._path, target)
+        status = os.system(cmd)
+        if status != 0:
+            raise TankGitError(
+                "Error executing git operation. The git command '%s' "
+                "returned error code %s." % (cmd, status)
+            )
+        log.debug("Git clone into '%s' successful." % target)
+
+        # clone worked ok! Now execute git commands on this repo
+        cwd = os.getcwd()
+        output = None
+        try:
+            log.debug("Setting cwd to '%s'" % target)
+            os.chdir(target)
+            for command in commands:
+
+                full_command = "git %s" % command
+                log.debug("Executing '%s'" % full_command)
+
+                try:
+                    output = subprocess_check_output(
+                        full_command,
+                        stderr=subprocess.STDOUT,
+                        shell=True
+                    )
+
+                    # note: it seems on windows, the result is sometimes wrapped in single quotes.
+                    output = output.strip().strip("'")
+
+                except SubprocessCalledProcessError, e:
+                    raise TankGitError(
+                        "Error executing git operation '%s': %s (Return code %s)" % (full_command, e.output, e.returncode)
+                    )
+                log.debug("Execution successful. stderr/stdout: '%s'" % output)
+
+        finally:
+            log.debug("Restoring cwd (to '%s')" % cwd)
+            os.chdir(cwd)
+
+        # return the last returned stdout/stderr
+        return output
```

Comment:
For the current use cases, no, it wouldn't make sense. In all cases, the pattern generally followed here is 1. establish state, 2. query for information. This is an internal method so i guess we can always make adjustments if future needs arise.